### PR TITLE
Implement Desktop Heartbeat Controller LogHeartbeat

### PR DESCRIPTION
**Pull request by AlmostMatt**
### Description
Defined heartbeat desktop controller.
Implements LogHeartbeat.
Adds tests to verify successful logging of heartbeats.
***
### Testing
Added unit tests.
Using the heartbeat desktop storage class directly to verify successsful logging. Using a mock date provider in order to test cases related to same date or older date.
***

### Type of Change
Place an `x` the applicable box:
- [ ] Bug fix. Add the issue # below if applicable.
- [x] New feature. A non-breaking change which adds functionality.
- [ ] Other, such as a build process or documentation change.
***

#### Notes
- Bug fixes and feature changes require an update to the `Release Notes` section of `release_build_files/readme.md`.
- Read the contribution guidelines [CONTRIBUTING.md](https://github.com/firebase/firebase-cpp-sdk/blob/main/CONTRIBUTING.md).
- Changes to the public API require an internal API review. If you'd like to help us make Firebase APIs better, please propose your change in a feature request so that we can discuss it together.

### CODE REVIEW

File: app/src/heartbeat/heartbeat_controller_desktop.cc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    if (logged_heartbeats.last_logged_date != current_date) {
+      logged_heartbeats.last_logged_date = current_date;
+      logged_heartbeats.heartbeats[user_agent].push_back(current_date);
+      // Don't store more than 30 days for the same user agent
+      if (logged_heartbeats.heartbeats[user_agent].size() > 30) {
+        logged_heartbeats.heartbeats[user_agent].erase(
+          logged_heartbeats.heartbeats[user_agent].begin());
+      }
+    }
+    bool write_succeeded = this->storage_.Write(logged_heartbeats);
```
Comment (chkuang-g) [inline app/src/heartbeat/heartbeat_controller_desktop.cc:63]: I think we need to have a way to lock the file after read and before write to prevent race. Here are two scenarios:

- In the same application instance, the user created two `FirebaseApp` with slightly different `AppConfig`, ex. database url, but still using the same `app_id`. In this case, there will be two schedular thread attempting to read/write the same file.
- The user launched two instances of the same application.

The race condition is:
- Instance/Thread A call `LogHeartbeat()` and read from disk
- Instance/Thread B call `GetAndFlushPayload()` and clear the file
- Instance/Thread A write all the log back to the file.

File: app/tests/heartbeat_controller_desktop_test.cc [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  // Even though heartbeat logging is asynchronous, it happens in the order that it is scheduled
+  // So the heartbeats should be logged in the correct order.
+  for (auto const& entry : read_heartbeats.heartbeats) {
+    std::vector<std::string> dates = entry.second;
+    ASSERT_EQ(dates.size(), 30);
+    EXPECT_EQ(dates[0], "2000-03-01");
+    EXPECT_EQ(dates[29], "2000-03-30");
+  }
+}
+
```
Comment (chkuang-g) [inline app/tests/heartbeat_controller_desktop_test.cc:220]: Is there plan to add more edge case here?
For instance:
- What if log `2000-01-25` then `2000-01-24`?
- Create destroy tests for race condition
- Having two LoggedHeartbeats instances pointing to the same app_id
etc.

**Review by chkuang-g — CHANGES_REQUESTED** [review]

**Review by chkuang-g — CHANGES_REQUESTED** [review]
Also
https://github.com/firebase/firebase-cpp-sdk/runs/6994613799?check_suite_focus=true

File: app/src/heartbeat/heartbeat_controller_desktop.cc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    // If last logged timestamp is not todays date, add a heartbeat
+    if (logged_heartbeats.last_logged_date != current_date) {
+      logged_heartbeats.last_logged_date = current_date;
+      logged_heartbeats.heartbeats[user_agent].push_back(current_date);
+      // Don't store more than 30 days for the same user agent
+      if (logged_heartbeats.heartbeats[user_agent].size() > 30) {
+        logged_heartbeats.heartbeats[user_agent].erase(
+          logged_heartbeats.heartbeats[user_agent].begin());
+      }
+    }
```
[the remark this answers is not in the document]
Comment (AlmostMatt) [reply app/src/heartbeat/heartbeat_controller_desktop.cc:69]: Done.

File: app/src/heartbeat/heartbeat_controller_desktop.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlmostMatt) [reply app/src/heartbeat/heartbeat_controller_desktop.cc:70]: Done. Restructured code so that the if block results in an early return.

File: app/src/heartbeat/heartbeat_controller_desktop.cc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+HeartbeatController::~HeartbeatController() {}
+
+void HeartbeatController::LogHeartbeat() {
+  std::string user_agent = App::GetUserAgent();
+
+  std::function<void(void)> log_heartbeat_funct = [&, user_agent]() {
+    std::string current_date = date_provider_.GetDate();
+    // Return early if a heartbeat has already been logged today.
+    if (this->last_logged_date_ == current_date) {
```
[the remark this answers is not in the document]
Comment (AlmostMatt) [reply app/src/heartbeat/heartbeat_controller_desktop.cc]: Changed to use <= and > instead of == and !=

File: app/src/heartbeat/heartbeat_controller_desktop.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  const std::string& app_id, const Logger& logger, const DateProvider& date_provider)
+  : storage_(app_id, logger), scheduler_(), last_logged_date_(""), date_provider_(date_provider)
+{}
+
+HeartbeatController::~HeartbeatController() {}
+
+void HeartbeatController::LogHeartbeat() {
+  std::string user_agent = App::GetUserAgent();
+
+  std::function<void(void)> log_heartbeat_funct = [&, user_agent]() {
```
[the remark this answers is not in the document]
Comment (AlmostMatt) [reply app/src/heartbeat/heartbeat_controller_desktop.cc]: Added a comment in the header before last_logged_date_

File: app/src/heartbeat/heartbeat_controller_desktop.cc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+HeartbeatController::HeartbeatController(
+  const std::string& app_id, const Logger& logger, const DateProvider& date_provider)
+  : storage_(app_id, logger), scheduler_(), last_logged_date_(""), date_provider_(date_provider)
+{}
+
+HeartbeatController::~HeartbeatController() {}
+
+void HeartbeatController::LogHeartbeat() {
+  std::string user_agent = App::GetUserAgent();
```
[the remark this answers is not in the document]
Comment (AlmostMatt) [reply app/src/heartbeat/heartbeat_controller_desktop.cc]: Done

File: app/tests/heartbeat_controller_desktop_test.cc (hunk 2 above)
Comment (AlmostMatt) [reply app/tests/heartbeat_controller_desktop_test.cc:233]: Added a test for logging dates in reverse-chronological order and verifying that only the newest date is logged.

Added a test that calls LogHeartbeat many times and then calls the controller's Destructor
- note: current implementation of Scheduler's destructor is to stop allowing scheduling of new work and to join with the worker thread. So all already-scheduled work is actually completed.

Added a test with multiple controllers for the same app id trying to log heartbeats simultaneously. Confirmed that it currently fails due to concurrency issues of read/write happening between other read/write operations, so the test is disabled.
Filed b/237003018 to track this

File: app/tests/heartbeat_controller_desktop_test.cc [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+TEST_F(HeartbeatControllerDesktopTest, DestroyControllerWhileWorkIsScheduled) {
+  HeartbeatController* destructible_controller =
+      new HeartbeatController(kAppId, logger_, mock_date_provider_);
+  {
+    // InSequence guarantees that all of the expected calls occur in order.
+    testing::InSequence seq;
+    for (int year = 2001; year <= 3000; year++) {
+      std::string date_string = std::to_string(year) + "-01-01";
+      EXPECT_CALL(mock_date_provider_, GetDate()).WillOnce(Return(date_string));
```
Comment (chkuang-g) [inline app/tests/heartbeat_controller_desktop_test.cc:276]: Hum. if you delete the controller right after scheduling those tasks, there is no guarantee that those tasks will run.
https://github.com/firebase/firebase-cpp-sdk/blob/b0a90622bf6ec231dd6a9fcae8702472f73d6d39/app/src/scheduler.cc#L72

Therefore, you cannot expect data_provider being called.  Otherwise, this test is very likely to fail occasionally.

I would recommend you to just call `LogHeartbeat()` once. But do Line 269-284 1000 times :)

File: app/tests/heartbeat_controller_desktop_test.cc [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    ASSERT_EQ(dates.size(), 30);
+    EXPECT_EQ(dates[0], "2000-03-01");
+    EXPECT_EQ(dates[29], "2000-03-30");
+  }
+}
+
+TEST_F(HeartbeatControllerDesktopTest, DestroyControllerWhileWorkIsScheduled) {
+  HeartbeatController* destructible_controller =
+      new HeartbeatController(kAppId, logger_, mock_date_provider_);
+  {
```
Comment (chkuang-g) [inline app/tests/heartbeat_controller_desktop_test.cc:271]: This bracket and indentation is not necessary

File: app/tests/heartbeat_controller_desktop_test.cc [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+  }
+  for (int i = 1; i <= 1000; i++) {
+    destructible_controller->LogHeartbeat();
+  }
+  // Trigger the controller's destructor before all of the async calls to
+  // LogHeartbeat have completed.
+  delete destructible_controller;
+
+  // Wait a bit to verify that no async threads seg fault
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```
Comment (chkuang-g) [inline app/tests/heartbeat_controller_desktop_test.cc:287]: Why do you need this and what is causing seg fault?

when you destroy destructible_controller, the thread created by the scheduler should be closed before the destructor ends.

**Review by chkuang-g — CHANGES_REQUESTED** [review]

File: app/tests/heartbeat_controller_desktop_test.cc (hunk 9 above)
Comment (chkuang-g) [reply app/tests/heartbeat_controller_desktop_test.cc:287]: If you need to wait to avoid seg fault, then this API is pretty unstable.

Comment (AlmostMatt) [reply app/tests/heartbeat_controller_desktop_test.cc:287]: The wait is not necessary, there is no seg fault, I have removed this.
I just thought it would be good to wait for async work to complete, but the scheduler destructor joins the worker thread so it already waits and this sleep is not relevant.

File: app/tests/heartbeat_controller_desktop_test.cc (hunk 8 above)
Comment (AlmostMatt) [reply app/tests/heartbeat_controller_desktop_test.cc:271]: Removed indendation

File: app/tests/heartbeat_controller_desktop_test.cc (hunk 7 above)
Comment (AlmostMatt) [reply app/tests/heartbeat_controller_desktop_test.cc:276]: In practice the current scheduler implementation completes all scheduled work as part of the destructor when it joins with the worker thread. So the date provider was called the expected number of times when I was testing it. 

I like your suggestion to create and destroy the controller within the loop. That should more consistently run into any timing edge cases if they exist.

File: app/src/heartbeat/heartbeat_controller_desktop.cc (hunk 1 above)
Comment (AlmostMatt) [reply app/src/heartbeat/heartbeat_controller_desktop.cc:72]: Added this comment.

**Review by chkuang-g — DISMISSED** [review]

**Comment by AlmostMatt** [comment]
Merged from main
